core/arch/atomic_exports.h

```diff
@@ -723,7 +723,254 @@ atomic_dec_becomes_zero(volatile int *var)
+                /* FIXME i#3544: test the assembly!*/          \
+                __asm__ __volatile__("fence iorw,iorw\n\t"     \
+                                     "sb %0, 0(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 1(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 2(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 3(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 4(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 5(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 6(%1)\n\t"        \
+                                     "srli %0, %0, 8\n\t"      \
+                                     "sb %0, 7(%1)\n\t"        \
+                                     :                         \
+                                     : "r"(value), "r"(target) \
+                                     : "memory");              \
+            } while (0)
+#        define ATOMIC_8BYTE_ALIGNED_WRITE(target, value, hot_patch) \
+            do {                                                     \
+                /* Not currently used to write code */               \
+                ASSERT_CURIOSITY(!hot_patch);                        \
+                __asm__ __volatile__("fence iorw,iorw\n\t"           \
+                                     "sd %0, 0(%1)\n\t"              \
+                                     :                               \
+                                     : "r"(value), "r"(target)       \
+                                     : "memory");                    \
+            } while (0)
+#        define ATOMIC_8BYTE_ALIGNED_READ(addr_src, addr_res)       \
+            do {                                                    \
+                __asm__ __volatile__("ld %0, 0(%1)\n\t"             \
+                                     "fence iorw,iorw"              \
+                                     :                              \
+                                     : "r"(addr_src), "r"(addr_res) \
+                                     : "memory");                   \
+            } while (0)
+
+#        define ADDI_suffix(sfx, var, val)                               \
+            do {                                                         \
+                /* Result width is handled by load/store instructions */ \
+                __asm__ __volatile__("l" sfx " t0, 0(%0)\n\t"            \
+                                     "add t0, t0, " #val "\n\t"          \
+                                     "fence iorw,iorw\n\t"               \
```

Comment:
How does the fence help make this atomic? Since other processors/threads can still modify the data before the updated value is written.

Reply:
I've re-worked this to use "Atomic instructions" ISA extension.